bubblewrap.c

```diff
@@ -450,33 +476,70 @@ do_init (int event_fd, pid_t initial_pid, struct sock_fprog *seccomp_prog)
   return initial_exit_status;
 }
 
-/* low 32bit caps needed */
-#define REQUIRED_CAPS_0 (CAP_TO_MASK (CAP_SYS_ADMIN) | CAP_TO_MASK (CAP_SYS_CHROOT) | CAP_TO_MASK (CAP_NET_ADMIN) | CAP_TO_MASK (CAP_SETUID) | CAP_TO_MASK (CAP_SETGID))
-/* high 32bit caps needed */
-#define REQUIRED_CAPS_1 0
-
 static void
-set_required_caps (void)
+set_required_caps (bool setup)
 {
   struct __user_cap_header_struct hdr = { _LINUX_CAPABILITY_VERSION_3, 0 };
   struct __user_cap_data_struct data[2] = { { 0 } };
 
   /* Drop all non-require capabilities */
-  data[0].effective = REQUIRED_CAPS_0;
-  data[0].permitted = REQUIRED_CAPS_0;
-  data[0].inheritable = 0;
-  data[1].effective = REQUIRED_CAPS_1;
-  data[1].permitted = REQUIRED_CAPS_1;
-  data[1].inheritable = 0;
+  data[0].effective = REQUIRED_SETUP_CAPS_0;
+  data[0].permitted = REQUIRED_SETUP_CAPS_0;
+  data[0].inheritable = setup ? 0 : requested_caps[0];
+  data[1].effective = REQUIRED_SETUP_CAPS_1;
+  data[1].permitted = REQUIRED_SETUP_CAPS_1;
+  data[1].inheritable = setup ? 0 : requested_caps[1];
   if (capset (&hdr, data) < 0)
```

Comment:
This will *always* run with all the possibly allowed caps in the setup. *even* in the setud root case. That seems bad, we're intentionally trying to run with minimal capabilities.

You only need to keep the capabilities that we really require for the setup (the old REQUIRED_CAPS_0), and then whatever the user requested.

In fact, I think we should only ever use the old REQUIRED_CAPS as effective, and only store the requested but not required-for-setup in "permitted".